src/main/scala/org/broadinstitute/dsde/firecloud/service/ExportEntitiesByTypeActor.scala

```diff
@@ -1,79 +1,208 @@
+      val errorReport = ErrorReport(StatusCodes.InternalServerError, s"Error generating entity download for $workspaceNamespace:$workspaceName:$entityType " + t.getMessage)
+      Future(ctx.complete(HttpResponse(
+        status = StatusCodes.InternalServerError,
+        entity = HttpEntity(ContentTypes.`application/json`, errorReport.toJson.compactPrint))))
   }
 
-  private def getZipBytes(entityType: String, membershipData: String, entityData: String): Array[Byte] = {
-    val bos = new ByteArrayOutputStream()
-    val zos = new ZipOutputStream(bos)
-    zos.putNextEntry(new ZipEntry(entityType + "_membership.tsv"))
-    zos.write(membershipData.getBytes)
-    zos.closeEntry()
-    zos.putNextEntry(new ZipEntry(entityType + "_entity.tsv"))
-    zos.write(entityData.getBytes)
-    zos.closeEntry()
-    zos.finish()
-    bos.toByteArray
+
+  /*
+   * Helper Methods
+   */
+
+  /**
+    * General Approach
+    * 1. Define a `Source` of entity queries
+    * 2. Run the source events through a `Flow`.
+    * 3. Flow sends events (batch of entities) to a streaming output actor
+    * 4. Return a Done to the calling route when complete.
+    */
+  private def streamSingularType(ctx: RequestContext, workspaceNamespace: String, workspaceName: String, entityType: String, entityQueries: Seq[EntityQuery], metadata: EntityTypeMetadata, headers: IndexedSeq[String], attributeNames: Option[IndexedSeq[String]]): Future[Done] = {
+    // The output to the user
+    val streamingActorRef = actorRefFactory.actorOf(Props(new StreamingActor(ctx, ContentTypes.`text/plain`, entityType + ".txt")))
+
+    // The Source
+    val entityQuerySource = Source(entityQueries.toStream)
+
+    // MapAsync should preserve order.
+    val flow = Flow[EntityQuery].mapAsync(1) { query =>
+      getEntities(workspaceNamespace, workspaceName, entityType, query) map { entities =>
+        sendRowsAsChunks(streamingActorRef, query, entityQueries.size, entityType, headers, entities)
+      }
+    }
+
+    // Ignore the result - we don't need to remember anything about this operation.
+    val sink = Sink.ignore
+
+    // finally, run it:
+    entityQuerySource.via(flow).runWith(sink)
+  }
+
+  private def streamCollectionType(ctx: RequestContext, workspaceNamespace: String, workspaceName: String, entityType: String, entityQueries: Seq[EntityQuery], metadata: EntityTypeMetadata, attributeNames: Option[IndexedSeq[String]]): Future[Done] = {
+    // The output file
+    lazy val zipFile = writeCollectionTypeZipFile(workspaceNamespace, workspaceName, entityType, entityQueries, metadata, attributeNames)
+
+    // The output to the user
+    lazy val streamingActorRef = actorRefFactory.actorOf(Props(new StreamingActor(ctx, ContentTypes.`application/octet-stream`, entityType + ".zip")))
+    zipFile map { f =>
+      streamingActorRef ! FirstChunk(HttpData.apply(f.byteArray), 0)
+    }
+
+    Future(Done)
```

Comment:
I spent a lot of time trying to make the chunks an `ask` instead of a `tell` and failed for reasons I don't understand. If I could get the ask pattern to work here, it would definitely clarify the upstream handling.